app/src/main/java/app/tivi/ViewModelStore.kt

```diff
@@ -0,0 +1,106 @@
+/*
+ * Copyright 2021 Google LLC
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package app.tivi
+
+import androidx.compose.runtime.Stable
+import kotlinx.coroutines.CoroutineScope
+import kotlinx.coroutines.Dispatchers
+import kotlinx.coroutines.Job
+import kotlinx.coroutines.SupervisorJob
+import kotlinx.coroutines.awaitCancellation
+import kotlinx.coroutines.cancel
+import kotlinx.coroutines.flow.SharingStarted
+import kotlinx.coroutines.flow.StateFlow
+import kotlinx.coroutines.flow.flow
+import kotlinx.coroutines.flow.onCompletion
+import kotlinx.coroutines.flow.stateIn
+import kotlinx.coroutines.isActive
+
+/**
+ * TODO
+ */
+@Stable
+internal class ViewModelStore(
+    coroutineScope: CoroutineScope,
+) {
+    private val viewModelMap = HashMap<Any, ViewModelStoreEntry>()
+
+    // Create a copy of the provided CoroutineScope, but replacing the Job with
+    // a child SupervisorJob. This allows a ViewModel to be cancelled without affecting any
+    // other ViewModels.
+    private val coroutineScope = CoroutineScope(
+        coroutineScope.coroutineContext +
+            SupervisorJob(coroutineScope.coroutineContext[Job])
+    )
+
+    /**
+     * Retrieve or create a ViewModel.
+     */
+    fun <T : Any> viewModelFlow(
+        key: Any,
+        create: (CoroutineScope) -> T
+    ): StateFlow<T> {
+        // First check if we have a 'cached' ViewModel which is still active
+        val cached = viewModelMap[key]?.takeIf { it.isActive }
+        if (cached != null) {
+            @Suppress("UNCHECKED_CAST")
+            return cached.flow as StateFlow<T>
+        }
+
+        // If not we'll create a new ViewModel
+
+        val viewModelScope = createChildCoroutineScope()
+
+        val flow = flow<T> {
+            // This is a 'fake' flow which doesn't emit anything. The actual ViewModel is
+            // created in the stateIn() below. We suspend the flow until its cancelled, which
+            // happens when either the stateIn `WhileSubscribed`, or the parent scope is cancelled
+            awaitCancellation()
+        }.onCompletion {
+            // The flow has been cancelled, so we need to also cancel the
+            // ViewModel's coroutine scope
+            viewModelScope.cancel()
+        }.stateIn(
+            scope = coroutineScope,
+            started = SharingStarted.WhileSubscribed(5000),
```

Comment:
This should be `SharingStartedWhileSubscribed(5000, replayExpirationMillis = 0)`.

Once `viewModelScope.cancel()` is called, the ViewModel is unusable, we'd need to create a new instance of it. However, `replayExpirationMillis` defaults to `Long.MAX_VALUE` which means that the same instance of the ViewModel will be _always_ used. We don't want that as its `coroutineScope` will be already cancelled. 

Ideally, we'd create a whole new ViewModel with a new `viewModelScope`.